From 8de24db5de6063feb1d23f9274fb52feb5995714 Mon Sep 17 00:00:00 2001
From: Yu Watanabe <watanabe.yu+github@gmail.com>
Date: Tue, 14 Apr 2026 08:57:47 +0900
Subject: [PATCH] sd-dhcp-client: do not trigger assertion when running
 on interface with small MTU

This also drops spurious default value. If mtu is unset, now the correct
default value is set in client_new_message() in dhcp-client-send.c.

[Adapted to 257.13: 257.13 has no dhcp-client-send.c, and the MTU feeds
only the Maximum Message Size option, so the default of
DHCP_MIN_PACKET_SIZE stays, and that option is computed from the MTU
bounded below by DHCP_MIN_PACKET_SIZE, as upstream bounds it by
IPV4_MIN_REASSEMBLY_SIZE.]

https://github.com/systemd/systemd/commit/8de24db5de6063feb1d23f9274fb52feb5995714
(cherry picked from commit 8de24db5de6063feb1d23f9274fb52feb5995714)
---
 src/libsystemd-network/sd-dhcp-client.c | 8 +++++---
 1 file changed, 5 insertions(+), 3 deletions(-)

diff --git a/src/libsystemd-network/sd-dhcp-client.c b/src/libsystemd-network/sd-dhcp-client.c
index c5a4cbe580..5fa30958cf 100644
--- a/src/libsystemd-network/sd-dhcp-client.c
+++ b/src/libsystemd-network/sd-dhcp-client.c
@@ -561,13 +561,14 @@ int sd_dhcp_client_set_port(
 
 int sd_dhcp_client_set_mtu(sd_dhcp_client *client, uint32_t mtu) {
         assert_return(client, -EINVAL);
-        assert_return(mtu >= DHCP_MIN_PACKET_SIZE, -ERANGE);
 
         /* MTU may be changed by the acquired lease. Hence, we cannot require that the client is stopped here.
          * Please do not add assertion for !sd_dhcp_client_is_running(client) here. */
 
-        client->mtu = mtu;
+        if (mtu < IPV4_MIN_MTU)
+                return -ERANGE;
 
+        client->mtu = mtu;
         return 0;
 }
 
@@ -905,7 +906,8 @@ static int client_message_init(
         /* RFC7844 section 3:
            SHOULD NOT contain any other option. */
         if (!client->anonymize && IN_SET(type, DHCP_DISCOVER, DHCP_REQUEST)) {
-                be16_t max_size = htobe16(MIN(client->mtu - DHCP_IP_UDP_SIZE, (uint32_t) UINT16_MAX));
+                be16_t max_size = htobe16(MIN(MAX(client->mtu, (uint32_t) DHCP_MIN_PACKET_SIZE) - DHCP_IP_UDP_SIZE,
+                                              (uint32_t) UINT16_MAX));
                 r = dhcp_option_append(&packet->dhcp, optlen, &optoffset, 0,
                                        SD_DHCP_OPTION_MAXIMUM_MESSAGE_SIZE,
                                        2, &max_size);
